# Проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. 
В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


Содержание.

1. <a href='#1'>Посчитайте, сколько книг вышло после 1 января 2000 года. </a><br/>


2. <a href='#2'>Для каждой книги посчитайте количество обзоров и среднюю оценку. </a><br/>


3. <a href='#3'>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.</a><br/>


4. <a href='#4'>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.</a><br/>


5. <a href='#5'>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a><br/>

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

In [3]:
sql = ''' select * from books'''
select(sql)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
sql = ''' select * from authors'''
select(sql)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [5]:
sql = ''' select * from ratings'''
select(sql)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [6]:
sql = ''' select * from reviews'''
select(sql)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [7]:
sql = ''' select * from publishers'''
select(sql)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

### 1. Посчитайте, сколько книг вышло после 1 января 2000 года. <a id='1'></a>

In [8]:
sql = ''' select count(t.book_id) from books t
where t.publication_date>'2000-01-01'
'''
select(sql)

count
0    819

Почти все книги из представленных выпущены после 2000 года

### 2. Для каждой книги посчитайте количество обзоров и среднюю оценку. <a id='2'></a>

In [9]:
sql = ''' 

select t.book_id, t.title,   count(distinct t.review_id) as cnt, avg(t.rating) as mean_rating 
from
    (select b.*, r.rating, re.review_id

    from books b
    left join ratings r on r.book_id=b.book_id
    left join reviews re on re.book_id=b.book_id
    ) t
group by 1, 2
order by 1,2

'''
select(sql)

book_id                                              title  cnt  \
0          1                                       'Salem's Lot    2   
1          2                 1 000 Places to See Before You Die    1   
2          3  13 Little Blue Envelopes (Little Blue Envelope...    3   
3          4  1491: New Revelations of the Americas Before C...    2   
4          5                                               1776    4   
..       ...                                                ...  ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)    3   
996      997                        Xenocide (Ender's Saga  #3)    3   
997      998                                    Year of Wonders    4   
998      999                        You Suck (A Love Story  #2)    2   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...    4   

     mean_rating  
0       3.666667  
1       2.500000  
2       4.666667  
3       4.500000  
4       4.000000  
..           ...  
995     3.666667  
996     3.400000  
997     3.200000  
998     4.500000  
999     3.833333  

[1000 rows x 4 columns]

### 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры. <a id='3'></a>

In [10]:
sql = ''' 
select t.* from
    (select p.publisher_id, p.publisher, count(distinct b.book_id) as cnt --, count(b.title)

    from books b
    left join publishers p on p.publisher_id=b.publisher_id
    where b.num_pages>50
    group by 1,2
    order by count(b.book_id) desc
    ) t
where t.cnt = (select max(t.cnt) from 
    (select p.publisher_id, p.publisher, count(distinct b.book_id) as cnt --, count(b.title)

    from books b
    left join publishers p on p.publisher_id=b.publisher_id
    where b.num_pages>50
    group by 1,2
    order by count(b.book_id) desc
) t)

'''
select(sql)

publisher_id      publisher  cnt
0           212  Penguin Books   42

Лидер - издание Penguin Books, выпустил 42 книги за все время наблюдений

### 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками. <a id='4'></a>

In [11]:
sql = ''' 
select t.*
from
(select a.author, avg(r.rating) as mean_rating, count(*) as cnt_rating
    from books b
    left join authors a on a.author_id=b.author_id
    left join ratings r on r.book_id=b.book_id
    
group by 1
) t
where t.cnt_rating>50 and t.mean_rating = (select max(t.mean_rating) from
(select t.*
from
(select a.author, avg(r.rating) as mean_rating, count(*) as cnt_rating
    from books b
    left join authors a on a.author_id=b.author_id
    left join ratings r on r.book_id=b.book_id
    
group by 1
) t
where t.cnt_rating>50
) t)
'''
select(sql)

author  mean_rating  cnt_rating
0  J.K. Rowling/Mary GrandPré     4.288462         312

Автор J.K. Rowling/Mary GrandPré выставил 312 оценки и средняя оценка 4.29 - самый высший результат среди всех

### 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок. <a id='5'></a>

In [12]:
sql = '''
select avg(t.count) from (
select t.username, count(distinct t.review_id) from (
select distinct t.username, re.review_id from (
select r.username, count(distinct r.rating_id) as cnt_rating
from ratings r
group by 1
) t
left join reviews re on re.username=t.username
where t.cnt_rating > 50
) t
group by 1
) t
'''
select(sql)

avg
0  24.333333

В среднем самые активные обзорщики выставляют 24 обзора